ingenialink/servo.py

```diff
@@ -195,123 +188,1052 @@ def stop(self):
+                               )
+            except ILError as e:
+                logger.error("Exception during load_configuration, register "
+                             "%s: %s", str(element.attrib['id']), e)
 
-    @abstractmethod
     def save_configuration(self, config_file, subnode=None):
-        raise NotImplementedError
+        """Read all dictionary registers content and put it to the dictionary
+        storage.
 
-    @abstractmethod
-    def store_parameters(self, subnode=None):
-        raise NotImplementedError
+        Args:
+            config_file (str): Destination path for the configuration file.
+            subnode (int): Subnode of the axis.
+
+        """
+        if subnode is not None and (not isinstance(subnode, int) or subnode < 0):
+            raise ILError('Invalid subnode')
+        prod_code, rev_number = get_drive_identification(self, subnode)
+
+        with open(self._dictionary.path, 'r', encoding='utf-8') as xml_file:
+            tree = ET.parse(xml_file)
+        root = tree.getroot()
+
+        body = root.find('Body')
+        device = root.find('Body/Device')
+        categories = root.find('Body/Device/Categories')
+        errors = root.find('Body/Errors')
+
+        if 'ProductCode' in device.attrib and prod_code is not None:
+            device.attrib['ProductCode'] = str(prod_code)
+        if 'RevisionNumber' in device.attrib and rev_number is not None:
+            device.attrib['RevisionNumber'] = str(rev_number)
+
+        registers_category = root.find('Body/Device/Registers')
+        if registers_category is None:
+            # Multiaxis dictionary
+            axes_category = root.find('Body/Device/Axes')
+            list_axis = root.findall('Body/Device/Axes/Axis')
+            self.__update_multiaxis_dict(device, axes_category, list_axis, subnode)
+        else:
+            # Single axis dictionary
+            registers = root.findall('Body/Device/Registers/Register')
+            self.__update_single_axis_dict(registers_category, registers, subnode)
+
+        device.remove(categories)
+        body.remove(errors)
+
+        image = root.find('./DriveImage')
+        if image is not None:
+            root.remove(image)
+
+        tree.write(config_file)
+        xml_file.close()
 
-    @abstractmethod
     def restore_parameters(self, subnode=None):
-        raise NotImplementedError
+        """Restore all the current parameters of all the slave to default.
+
+        .. note::
+            The drive needs a power cycle after this
+            in order for the changes to be properly applied.
+
+        Args:
+            subnode (int): Subnode of the axis. `None` by default which restores
+            all the parameters.
+
+        Raises:
+            ILError: Invalid subnode.
+            ILObjectNotExist: Failed to write to the registers.
+
+        """
+        if subnode is None:
+            # Restore all
+            self.write(reg=self.RESTORE_COCO_ALL,
```

Comment:
Function write and property RESTORE_COCO_ALL don't exist in Servo class. Functions should be added as abstract functions and properties as abstract properties (https://stackoverflow.com/questions/5960337/how-to-create-abstract-properties-in-python-abstract-classes) or defined as None in the parent (abstract properties seems overkill for me, what do you think?)

Reply:
Done